File: backend/repositories/user_repository.py
from extensions import db
from models.user import User


class UserRepository:
    def find_by_id(self, user_id: int) -> User | None:
        return db.session.get(User, user_id)

    def find_by_username(self, username: str) -> User | None:
        return db.session.execute(
            db.select(User).where(User.username == username)
        ).scalar_one_or_none()

    def find_by_username_and_role(self, username: str, role: str) -> User | None:
        return db.session.execute(
            db.select(User).where(User.username == username, User.role == role)
        ).scalar_one_or_none()

    def create(self, username: str, password: str, full_name: str, role: str, dob=None) -> User:
        user = User(username=username, password=password, full_name=full_name, role=role, dob=dob)
        db.session.add(user)
        db.session.flush()
        return user

    def update_password(self, user_id: int, hashed: str) -> None:
        user = db.session.get(User, user_id)
        if user:
            user.password = hashed

    def update_full_name(self, user_id: int, full_name: str) -> None:
        user = db.session.get(User, user_id)
        if user:
            user.full_name = full_name

    def search_students(self, q: str, limit: int = 20) -> list:
        from sqlalchemy import text
        sql = text("""
            SELECT u.id, u.username, u.full_name, u.dob, u.class_id, c.class_name AS current_class
            FROM users u
            LEFT JOIN classes c ON c.id = u.class_id
            WHERE u.role = 'student' AND u.username LIKE :q
            LIMIT :lim
        """)
        rows = db.session.execute(sql, {"q": f"%{q}%", "lim": limit}).mappings().all()
        return [dict(r) for r in rows]

File: backend/src/repositories/user.repository.js
import { query, queryOne, insert } from '../config/database.js';

export const findById = (id) =>
  queryOne('SELECT * FROM users WHERE id = :id', { id });

export const findByUsername = (username) =>
  queryOne('SELECT * FROM users WHERE username = :username', { username });

export const findByUsernameAndRole = (username, role) =>
  queryOne('SELECT * FROM users WHERE username = :username AND role = :role', { username, role });

export const create = async (username, password, fullName, role, dob = null) => {
  const id = await insert(
    'INSERT INTO users (username, password, full_name, role, dob) VALUES (:username, :password, :fullName, :role, :dob)',
    { username, password, fullName, role, dob }
  );
  return { id, username, full_name: fullName, role, dob };
};

export const updatePassword = (userId, hashed) =>
  query('UPDATE users SET password = :hashed WHERE id = :id', { hashed, id: userId });

export const getProfile = (userId) =>
  queryOne(
    'SELECT username, full_name, dob, email, phone FROM users WHERE id = :id',
    { id: userId }
  );

export const updateProfile = (userId, fullName, dob, email, phone) =>
  query(
    'UPDATE users SET full_name = :fullName, dob = :dob, email = :email, phone = :phone WHERE id = :id',
    { fullName, dob: dob || null, email: email || null, phone: phone || null, id: userId }
  );

export const searchStudents = (q, limit = 20) =>
  query(
    `SELECT u.id, u.username, u.full_name, u.dob, u.class_id, c.class_name AS current_class
     FROM users u
     LEFT JOIN classes c ON c.id = u.class_id
     WHERE u.role = 'student' AND u.username LIKE :q
     LIMIT :lim`,
    { q: `%${q}%`, lim: limit }
  );
